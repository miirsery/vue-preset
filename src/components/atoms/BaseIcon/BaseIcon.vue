<template>
  <svg class="base-icon" :class="iconClass" :width="props.width" :height="props.height" aria-hidden="true">
    <use :href="symbolId" />
  </svg>
</template>

<script setup lang="ts">
import { computed } from 'vue'

type Props = {
  name: string
  prefix?: string
  width?: string | number
  height?: string | number
  reverse?: boolean
}

const props = withDefaults(defineProps<Props>(), {
  prefix: 'icon',
  width: '23px',
  height: '24px',
  reverse: false,
})

const symbolId = computed(() => `#${props.prefix}-${props.name}`)

const iconClass = computed(() => {
  return [{ 'reversed-icon': props.reverse }, `base-icon--${props.name}`]
})
</script>

<style lang="scss" scoped>
.base-icon {
  position: relative;
  display: inline-block;
  flex-shrink: 0;
  transition: transform 0.2s;
}

.reversed-icon {
  transform: rotate(180deg);
}
</style>
