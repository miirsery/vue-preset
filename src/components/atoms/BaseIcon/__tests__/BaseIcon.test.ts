import { describe, expect, test } from 'vitest'
import { mount, VueWrapper } from '@vue/test-utils'

import BaseIcon from '@/components/atoms/BaseIcon/BaseIcon.vue'
import { nextTick } from 'vue'

// TODO: Добавить нормальную типизацию

const _mount = (options?: Record<string, unknown>): VueWrapper => {
  return mount(BaseIcon, {
    props: {
      name: 'close',
    },
    ...options,
  })
}

describe('Testing BaseIcon.vue', () => {
  test('Should render base icon', async () => {
    const wrapper = _mount()

    expect(wrapper.exists()).toBeTruthy()

    expect(wrapper.find('svg').exists()).toBeTruthy()
  })

  test('Should have snapshot', async () => {
    const wrapper = _mount()

    expect(wrapper.element).toMatchSnapshot()
  })

  test('Should have needed class', async () => {
    expect.arrayContaining(['base-icon', 'base-icon--close'])
  })

  test('Should change icon size', async () => {
    const wrapper = _mount()

    expect(wrapper.attributes().width).toContain('23px')

    expect(wrapper.attributes().height).toContain('24px')

    await wrapper.setProps({
      width: 48,
      height: 48,
    })

    await nextTick()

    expect(wrapper.attributes().width).toContain('48')

    expect(wrapper.attributes().height).toContain('48')

    await wrapper.setProps({
      width: '52px',
      height: '52px',
    })

    expect(wrapper.attributes().width).toContain('52px')

    expect(wrapper.attributes().height).toContain('52px')
  })

  test('Should be reversed', async () => {
    const wrapper = _mount()

    await wrapper.setProps({
      reverse: true,
    })

    await nextTick()

    expect(wrapper.classes()).toEqual(['base-icon', 'reversed-icon', 'base-icon--close'])
  })
})
