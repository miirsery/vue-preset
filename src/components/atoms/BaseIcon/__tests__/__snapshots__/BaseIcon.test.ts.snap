// Vitest Snapshot v1

exports[`Testing BaseIcon.vue > Should have snapshot 1`] = `
<svg
  aria-hidden="true"
  class="base-icon base-icon--close"
  data-v-8c85aa7d=""
  height="24px"
  width="23px"
>
  <use
    data-v-8c85aa7d=""
    href="#icon-close"
  />
</svg>
`;
